semantic-python/src/Language/Python/Failure.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE DeriveFoldable, DeriveFunctor, DeriveGeneric, DeriveTraversable, ExistentialQuantification,
+             FlexibleContexts, KindSignatures, MultiParamTypeClasses, QuantifiedConstraints, RankNTypes,
+             StandaloneDeriving, TypeOperators #-}
+
+module Language.Python.Failure
+  ( Failure (..)
+  , unimplemented
+  , invariantViolated
+  , eliminateFailures
+  ) where
+
+import Prelude hiding (fail)
+
+import Control.Effect.Carrier
+import Control.Monad.Fail
+import Data.Coerce
+import Data.Kind
+import GHC.Generics (Generic1)
+import Syntax.Foldable
+import Syntax.Module
+import Syntax.Term
+import Syntax.Traversable
+
+data Failure (f :: Type -> Type) a
+  = Unimplemented String
+  | InvariantViolated String
+    deriving Generic1
```

Comment:
Fun thought: once we’re ported to `fused-effects` 1.0, we can use `Control.Effect.Throw` + a simple, first-order error datatype with these constructors.